paddle/framework/framework.md

```diff
@@ -0,0 +1,205 @@
+      estimated = true)             # will be updated by the backward algorithm.
+x.estimated = false                 # prevents from update.
+```
+
+Note that the variable has the following methods:
+
+```cpp
+class Variable {
+ public:
+  bool Estimated() const;
+  bool SetEstimated(bool);
+  template <typename T> const T& Get() const;
+  template <typename T> T* GetMutable();
+  template <typename T> bool IsType(T) const;
+};
+```
+
+`Get` and `GetMutable` implements *lazy memory allocation*, as described in the [Variable design doc](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/framework/variable.md).
+
+Note that `name` is not a property of a variable.  A variable can have various names in different scopes.
+
+### Scope
+
+In programming languages, variables belong to scopes.  Scopes enable the release of local variables when a stack frame pops.
+
+A neural network is arguably equivalent to a program.  The *recurrent operator*, `paddle::operator::Recurrent`, is like `for` loop, and the conditional operator, `paddle::operator::Conditional`, is like `if/switch`.  They can have sub-networks as attributes, and `paddle::operator::Recurrent/Conditional::Run` creates a local scope before executing the sub-network.  At inference time, once `paddle::operator::Recurrent/Conditional::Run` frees the scope before completes.  At training time, it is the corresponding gradient operations' `paddle::operator::RecurrentGrad/ConditionalGrad::Run` who free the local scope.
+
+The global and nested local scopes form a hierarchy.  The following Python functions make it convenient to program scopes:
+
+1. `paddle.scope.current()` returns the current scope, which defaults to
+1. `paddle.scope.global()`, which returns the top-level scope.
+
+C++ code shouldn't maintain global status like the current scope to prevent unexpected inconsistancy.
+
+```cpp
+class Scope {
+ public:
+  Scope() : parent_(nullptr) {} // Constructor creates only global scopes. 
+  
+  Variable* FindVar(std::string name); // Find in the hierrachy or return null.
+  Variable* CreateVar(StringPiece name, Variable*); // Find or create.
+  
+  Scope* CreateScope(StringPiece name); // Finds or creates a sub-scope.
+  void DeleteScope(Scope*);             // Delete a sub-scope or raise exception
+  
+ private:
+  std::map<std::string /*name*/, std::unique_ptr<Variable> > vars_;
+  std::shared_ptr<Scope> parent_;
+  std::vector<Scope*> children_;
+  
+  Mutex mutex_;  // Make this class thread-safe.
+};
+```
+
+## Execution and Context
+
+A neural network is a program.  Training or inference is to execute it.  The runtime environment of execution is known as a *context*:
+
+1. a scope,
+1. device(s), or places,
+1. a flag indicates if we are training and should we create gradient operations and run backward
+
+and can be defined as
+
+```cpp
+struct Context {
+  Scope* scope_;
+  std::vector<Place> places_; // a network might run on multiple devices.
+  bool training_;
+};
+```
```

Comment:
Up to what I understand now, I'd avoid running a network on multiple devices, because the maintenance of multiple CUDA streams means a very complicated CUDAContext or OpKernelContext or something like that.

And it seems that we can make use of multiple devices using data parallelism, which requires only running a net on a single device.